### DISCUSSION TITLE
Button functionality

### DISCUSSION THREAD

**Discussion by trip5** [EdenRidgway/XY-Wifi-Clock#7] [Ideas]
Since it's on the roadmap, can we talk about what the buttons could do?

As I recall, the buttons with the Sinilink firmware are very typical of an alarm clock, one mode button and +/- set the time or alarm or brightness...  But with this firmware, we can leave all that to the webUI...

Personally, I'd like to see one of the buttons allow access to hotspot, time-limited to 5 minutes or so... to allow it to be used as a travel clock...  Namely because I tried it at my workplace where I know mdns doesn't work... and I realized that there was no way to retrieve the IP address.

I have no idea about the others... maybe a snooze for the alarm clock?  Maybe a display on/off switch?  What do you guys think?

--- Below this line is some to-do stuff for future, will edit as needed:

Snooze for alarm (Up or Down can do that).

**Comment by EdenRidgway** [comment]
Your thoughts on the button acting as a hotspot toggle and snooze make sense.

In terms of the IP address, one thought I had was that when it started it up it could scroll the IP address across the display (or maybe on a button combination) in that way one would not have to rely on MDNS. We need to sort out that pesky point display issue which would also help with the AM/PM time display.

**Reply by trip5** [reply]
I read the docs for the TM1650 but it really looks like it's colon OR dots... so... I think a pm-dot is out, but it may be OK to switch modes while showing the IP address...?  I don't think the original firmware had a pm-dot either.

I played with keeping the hotspot open but... when viewing on mobile, the fields stay unpopulated, sometimes reloading (multiple times) helps but it's hit-or-miss.  No problem if accessing the device through the wi-fi network.

It seems a common problem with serving web pages from ESP to Android, too: https://github.com/espressif/arduino-esp32/issues/4423 - in that case it's ESP32 and there's a weird workaround... but this is ESP8266 and after some digging, not sure if it's even related.  Espressif's docs seem to indicate that AMPDU is not enabled by default on ESP8266 and anyways lol, this isn't an ESP8266 anyways, right?  So...

I'm starting to think that showing the IP address is the best bet...

**Reply by EdenRidgway** [reply]
I've just pushed a change where it displays the IP address on startup and the dot underneath the colons when it is AM and in 12 hour mode. It would be good if we can get those buttons working properly as they seem to stay triggered after they are pressed. Maybe a library like EzButton or similar would help but I'm not sure (the solution may be a lot simpler than that).

**Reply by trip5** [reply]
Lol.  Should have read your note fully.  I really struggled trying to get my own thing going... instead I ended up confused trying to figure out why the buttons weren't de-bouncing.  I'll make that the next thing I look at.

Anyways, temporary hotspot code in... but I couldn't figure out an easy way to implement something like epoch time or now() to get the timeout to actually work if it crosses midnight.

Also, a dot for am?  Why not pm?  The code looks neat and the dot placement is lucky... the way the dots and colons are tied into the digits is so weird.  I'm just glad it didn't end up between 1&2 or 3&4.

**Reply by EdenRidgway** [reply]
The way the library ties the dot to the digit is weird. The only reason I made the dot represent AM is because it is to the left of the colon. I originally had hoped it could be setup with the left most dot being AM and the right most dot being PM, but as you are aware unfortunately the 3rd and 4th digits are tied to the colon (at least in the RM16xx library). 

More than happy for it to represent PM too. I think a single dot to represent either AM or PM is good enough for now as I can see a fair amount of time being spent trying to get it to work differently which might better be directed towards other features.

The way the buttons are working on the device is quite odd/unexpected and it makes me wonder if we need a different approach. It'll probably be one of those face palm moments when we finally realise what the solution is, or maybe it simply is just a strange hardware implementation.

**Reply by trip5** [reply]
I saw that ezbutton library. I put the code in.  It works. But there wasn't an easy way to handle long-presses, so I found Button2 https://github.com/LennartHennigs/Button2/ and it can handle even how long you pressed a button for... which could add a lot of functionality.

So... I put some demos in there but I already fixed the Set button for a long click to show the IP... holding it for 3s will activate (or deactivate) the Hotspot.  But there's something seriously wrong with my logic at this point...  so it's commented out for now.   Might be a lack of sleep.  I'll keep going tomorrow.  Lol.

**Comment by trip5** [comment]
The button functions are in but the Brightness adjustment appears not to work. I think it was commented out in some older code and I just put it back in as functions.  Anything the buttons do need to be functions because the Button2 code recognizes click and press differently...

I deactivated the show IP at boot.  It can be shown with a short click of the Set button instead.

**Reply by trip5** [reply]
On the topic of a buttons... and lol, I'm going to try to avoid doing this just yet... I need to get some Dying Light 2 time in for a few days.

Making a to-do list at top... Snooze seems like it's possible.  5 minutes?  7?

Any ideas for long-pressing the up or down buttons?
